### ISSUE TITLE
Migrate to FindCUDAToolkit.cmake

### ISSUE THREAD

**Issue by SomeoneSerge** [davisking/dlib#2833]
### Main idea

NVIDIA considers FindCUDA.cmake (`find_package(CUDA)`) deprecated: https://docs.nvidia.com/cuda/cuda-installation-guide-linux/#importing-tarballs-into-cmake. The common approach to consume cudatoolkit today is to use FindCUDAToolkit.cmake (`find_package(CUDAToolkit COMPONENTS cudart cublas ...)`).

Aside from FindCUDA being obsolete, there are notable differences between the modules: e.g. `FindCUDAToolkit.cmake` allows its `CUDAToolkit_ROOT` variable to take the form of a list of paths, and this way it supports splayed installations of cudatoolkit, where different components (cudart, cublas, ...) may reside in different prefixes (as is the case e.g. in [nixpkgs](https://github.com/NixOS/nixpkgs/))

Thanks!

**Comment by davisking** [comment]
Yeah, we will update eventually.  We are on cmake 3.8 right now and that's a feature only in 3.17 or newer.  Lots of dlib users update their tooling slowly, so dlib has to update slowly too :|

**Comment by jakirkham** [comment]
Would it make sense to have branch logic for these 2 cases? Namely

* If CMake is 3.17+, use `find_package(CUDAToolkit)`
* If CMake is older, fallback to `find_package(CUDA)`

Maybe another option would be to vendor [`FindCUDAToolkit.cmake`]( https://github.com/Kitware/CMake/blob/v3.28.1/Modules/FindCUDAToolkit.cmake )

cc @robertmaynard (in case you have thoughts/suggestions here)

**Comment by SomeoneSerge** [comment]
I like the idea of introducing FindCUDAToolkit.cmake support earlier. I find branching preferable to vendoring because it's a risk of divergence. E.g. pytorch vendors a copy, but their API already deviates from kitware's (the former doesn't accept lists in `CUDAToolkit_ROOT`). There are projects that went with branching too (list tbd)

**Comment by robertmaynard** [comment]
Replacing `FindCUDA` fully with `FindCUDAToolkit` will require also switching from the previous use of `cuda_add_library` ( and `CUDA_NVCC_FLAGS` ) to the first class CUDA language support in CMake ( `add_library(example source.cu)` ).

This kind of change makes it harder to support switching between pre and post 3.17. I would recommend making a clean break when possible, especially given that CMake 3.27 introduces a policy to remove `FindCUDA` entirely.

CMake does offer `apt.kitware.com` for ubuntu users that want to to get the latest version of CMake on older Ubuntu releases. The CMake binaries from `cmake.org` are built to support seriously old distro's as well, so they can be dropped onto machines which only provide 3.8 via the official package manager.

**Comment by Tobias-Fischer** [comment]
See https://github.com/conda-forge/dlib-feedstock/pull/83 for a patch that uses FindCUDAToolkit.cmake

**Comment by davisking** [comment]
Yeah if it can just branch on cmake version without being a big mess that would be cool.  I personally have newer cmake on my system.  Maybe we could start to require cmake 3.17 for dlib.  It bothers me though that, for example, Ubuntu 22.04 then wouldn't work with its version of cmake and I know plenty of people are still on stuff that old.

Is this a big deal to not have the newer thing right now?

**Comment by Tobias-Fischer** [comment]
CUDA12 is (as far as I know) not supported with the old FindCUDA.

**Comment by robertmaynard** [comment]
> Ubuntu 22.04 then wouldn't work with its version of cmake and I know plenty of people are still on stuff that old.

Ubuntu 22.04 provides CMake 3.22.
Ubuntu 20.04 provides CMake 3.16
Ubuntu 18.04 provides CMake 3.10

The upside is for 20.04 and 22.04 the `apt.kitware.com` repository can be used to get latest cmake releases ( officially blessed by CMake ). Ubuntu 18.04 users would need to download the pre-build cmake binaries and install them manually.

**Comment by davisking** [comment]
Uh, yeah, alright.  I've come around.  It's not like people have to pull the latest dlib if they are on an old OS anyway and it would be real convenient for multiple reasons to upgrade CMake.  So one of you want to put up a PR that switches to cleaner cuda cmake rules and uses cmake 3.17? :)

**Comment by Tobias-Fischer** [comment]
I don't have the capacity at the moment, but basically everything required is in https://github.com/conda-forge/dlib-feedstock/blob/cf9148326a7ad74c1dd519c2d3d5dc2912f9697c/recipe/use_new_cudatoolkit.patch (just needs a bump of the cmake_minimum_version).

**Comment by davisking** [comment]
Well, I mean, is it really all in that patch?  Seems like you are saying
it's not if it takes time? :)

**Comment by Tobias-Fischer** [comment]
Yes, it’s all in the patch except for setting the correct minimum cmake version

**Comment by pfeatherstone** [comment]
I did have a branch at some point in the past that cleaned up all of dlib's cmake stuff including CUDA. I can try to revive that at some point.

**Comment by davisking** [comment]
Yeah.  I think I'm fine with switching the required version of cmake to a higher version.  So if there is some good stuff to be had by doing that then send me a PR :)

**Comment by maxrudiy** [comment]
Hello. I'm not sure what is this topic about exactly. But I trying to install dlib on windows11 by cmake and with cuda support. And this is crazy because only way I could do it by manually moving cuda files like in this issue https://github.com/davisking/dlib/issues/2979 Looks like there are some mess with paths and all my paths are default

**Comment by Tobias-Fischer** [comment]
@maxrudiy - have you tried `conda install dlib -c conda-forge`?

**Comment by maxrudiy** [comment]
> [@maxrudiy](https://github.com/maxrudiy) - have you tried `conda install dlib -c conda-forge`?
Thanks for reply. I'm using venv and not sure whats wrong. There are no errors while setup "DLIB WILL USE CUDA, compute capabilities: 50". But then I have error "_dlib_pybind11: The specified module could not be found."
I tried to copy lib files from other pc (where i made it work yesterday somehow), but I have same error here with missing import. I don't know, could it be venv issue or some paths, or should i try conda. I'm using all latest versions, but python 3.11

**Comment by jakirkham** [comment]
What Tobias is suggesting is using [Conda]( https://conda.org ), which is a different installer. This may make it easier to install without the kinds of workarounds that you described

In any event, this is orthogonal of the discussion here. So if this needs more discussion, please raise a new issue

**Comment by smu-sc-gj** [comment]
Hi all, 

I think I ran into this myself yesterday, I had a try at fixing it before I checked the open issues and found this.  My fix didn't work, but I'll explain my failure here in the hope it will help. 

I'm running Fedora 42 and it's up to date 6.14.9-300.fc42.x86_64 with CUDA installed [based on this guide](https://forum.level1techs.com/t/cuda-12-9-on-fedora-42-guide-including-getting-cuda-samples-running/230769), I've also tested with the provided examples. 

Running CMake:

```bash
-- Found BLAS and LAPACK via pkg-config
CMake Warning (dev) at dlib/CMakeLists.txt:652 (find_package):
  Policy CMP0146 is not set: The FindCUDA module is removed.  Run "cmake
  --help-policy CMP0146" for policy details.  Use the cmake_policy command to
  set the policy and suppress this warning.

This warning is for project developers.  Use -Wno-dev to suppress it.

-- Found CUDA: /usr/local/cuda-12.9 (found suitable version "12.9", minimum required is "7.5")
-- Looking for cuDNN install...
-- *** cuDNN V5.0 OR GREATER NOT FOUND.                                                       ***
-- *** Dlib requires cuDNN V5.0 OR GREATER.  Since cuDNN is not found DLIB WILL NOT USE CUDA. ***
-- *** If you have cuDNN then set CMAKE_PREFIX_PATH to include cuDNN's folder.                ***
-- Disabling CUDA support for dlib.  DLIB WILL NOT USE CUDA
```

I tried to fix CMP0146, based on [this CUDA guide](https://cliutils.gitlab.io/modern-cmake/chapters/packages/CUDA.html).  I can't attach a .patch file so here's the content. 

```
--- ../dlib/CMakeLists.txt	2025-06-06 11:48:37.356453244 +0100
+++ ../../../dlib/dlib/CMakeLists.txt	2025-06-06 13:00:37.264892258 +0100
@@ -13,7 +13,8 @@
    cmake_policy(SET CMP0077 NEW)
 endif()
 
-project(dlib)
+# tried adding cuda here but we don't always need it.
+project(dlib C CXX)
 
 set(CPACK_PACKAGE_NAME "dlib")
 set(CPACK_PACKAGE_VERSION_MAJOR "20")
@@ -648,8 +649,24 @@
       endif()
 
 
-      if (DLIB_USE_CUDA)
-         find_package(CUDA 7.5)
+# hacking
+      include(CheckLanguage)
+      check_language(CUDA)
+      if (NOT CUDA)
+         set(CUDA_FOUND 0)
+      else()
+         set(CUDA_FOUND 1)
+      endif()
+
+# hacking
+      if (DLIB_USE_CUDA AND CUDA_FOUND)
+         #find_package(CUDA 7.5)
+         enable_language(CUDA)
+         if(NOT DEFINED CMAKE_CUDA_STANDARD)
+            set(CMAKE_CUDA_STANDARD 11)
+            set(CMAKE_CUDA_STANDARD_REQUIRED ON)
+         endif()
+         
 
          if (CUDA_VERSION VERSION_GREATER 9.1 AND CMAKE_VERSION VERSION_LESS 3.12.2)
             # This bit of weirdness is to work around a bug in cmake 
@@ -880,11 +897,17 @@
          # The old cuda_add_library() command doesn't support CMake's newer dependency
          # stuff, so we have to set the include path manually still, which we do here.
          include_directories(${dlib_needed_public_includes})
-         cuda_add_library(dlib ${source_files} )
+ # hacking
+         add_library(dlib ${source_files} )
+         #cuda_add_library(dlib ${source_files} )
+         set_target_properties(dlib PROPERTIES CUDA_SEPARABLE_COMPILATION ON)
+         set_target_properties(dlib PROPERTIES CUDA_ARCHITECTURES "35;50;72")
       else()
          add_library(dlib ${source_files} )
       endif()
 
+      
+
    endif ()  ##### end of if NOT DLIB_ISO_CPP_ONLY ##########################################################
```
This allows Dlib to compile and most of the examples, however, those which try to use CUDA compile but fail to link.  From the errors I'm guessing CUDA is not linked to the final executable?  

```
...
/usr/bin/ld: dlib_build/libdlib.a(tensor_tools.cpp.o): in function `dlib::tt::embeddings_gradient(dlib::tensor const&, dlib::tensor const&, dlib::tensor&, dlib::tensor const&, float, bool)':
/home/user/code/Dlib/dlib/dlib/cuda/tensor_tools.cpp:1409:(.text+0x3a97): undefined reference to `dlib::cuda::embeddings_gradient(dlib::tensor const&, dlib::tensor const&, dlib::tensor&, dlib::tensor const&, float, bool)'
...
```
